### ISSUE TITLE
Potential listener LEAK detected

### ISSUE THREAD

**Issue by alexandruavadanii** [microsoft/vscode-cpptools#12052]
Type: <b>Bug</b>

1. Create a new workspace with 3 empty files (`1.c`, `2.c`, `3.c`)
2. Open all 3 empty files in 3 columns (not relevant, but makes the next step easier)
3. Click a couple of times inside each of the 3 editors (1, 2, 3, 1, 2, 3 etc.)
4. Toggle developer console and analyze the logs
5. :bug: potential listener LEAK detected, having 439 listeners already.

![image](https://github.com/microsoft/vscode-cpptools/assets/11544783/523ac295-b61c-4035-b698-0a448822ef59)

You can probably ignore the "Redundant folding ranges request received..." for this, just filter out "info" messages from the developer console.

Extension version: 1.19.4
VS Code version: Code - Insiders 1.88.0-insider (61187a0b6547df5911a0ee259f7afda46bac7ec3, 2024-03-04T05:46:35.204Z)
OS version: Windows_NT x64 10.0.18362
Modes:

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i7-8665U CPU @ 1.90GHz (8 x 2112)|
|GPU Status|2d_canvas: enabled<br>canvas_oop_rasterization: enabled_on<br>direct_rendering_display_compositor: disabled_off_ok<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>opengl: enabled_on<br>rasterization: enabled<br>raw_draw: disabled_off_ok<br>skia_graphite: disabled_off<br>video_decode: enabled<br>video_encode: enabled<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled<br>webgpu: enabled|
|Load (avg)|undefined|
|Memory (System)|15.81GB (4.23GB free)|
|Process Argv||
|Screen Reader|no|
|VM|0%|

</details>

**Comment by alexandruavadanii** [comment]
Hi, @Colengms,
I think you got it right that the folding ranges are causing this.
I tried to bisect which vscode commit changed the behavior, but after seeing Martin's comments in https://github.com/microsoft/vscode/issues/206841, I think it's safe to assume this is caused by the doubled folding range requests coming in quick succession to vscode-cpptools. The second folding request was introduced with the sticky scroll feature, so that's why we don't see this with older vscode.

I will have to test this theory, but one possible explanation is a race condition between
https://github.com/microsoft/vscode-cpptools/blob/e12c7c28d15f6164117e38b4d61d68a20af67cda/Extension/src/LanguageServer/Providers/foldingRangeProvider.ts#L34 and
https://github.com/microsoft/vscode-cpptools/blob/v1.19.5/Extension/src/LanguageServer/Providers/foldingRangeProvider.ts#L45
- Request A for a document URI arrives.
- The code checks `pendingRequests` and finds no existing request for the URI.
- Request A initiates a language server request but hasn't yet added its `ManualPromise` to `pendingRequests`.
- Request B for the same URI arrives before step 3 completes.
- The code checks pendingRequests again and still finds no existing request for the URI, because Request A has not completed its registration.
- Request B initiates another language server request, leading to what we see.

I made a quick test with a home-brew mutex implementation guarding that section and both folding ranges info messages + potential leak detected warnings are gone.
